### DISCUSSION TITLE
get sender name and mailing address when downloading attachments

### DISCUSSION THREAD

**Discussion by bamper197801** [ecederstrand/exchangelib#1144] [Q&A] [answered]
How to get sender name and mailing address when downloading attachments.

**Comment by ecederstrand** [comment]
```python
from exchangelib import Message
Message.FIELDS
```

will show you all fields available on messages. You're probably looking for the `author` or `sender` fields.

**Comment by bamper197801** [comment]
Thanks for the answer.
I did not understand how to use your answer, but I used item.sender.name ,item.sender.email_address

I also wanted to ask how you can download attachments only from new letters, right now I have it set up by reverse sorting by the date of the letter and if there are no new letters, then the last one is always downloaded, is it possible by the dates of the letters, i.e. refer only to the new letter.

**Reply by ecederstrand** [reply]
There are a couple of options:

1. Use the `is_read` property to only process unread emails. When you have processed an email, mark the email as read by setting `is_read = True` and saving the email.
2. Fetch all emails. When you ahve processed an email, move it to a different folder.
3. Store the timestamp of the last time you processed entries, and filter emails after that timestamp on the next run.

**Comment by bamper197801** [comment]
I chose option 3, but there were difficulties.
saving the date-time of the last letter to a file, then saving it to a variable from the file, this variable is not visible in the loop and the condition does not work. where am I wrong? Thanks.

```python
#file_datetime_received = None
with open(path_attch +"\datetime_received.txt") as file:
    file_datetime_received = file.read()
#print(file_datetime_received)

for item in account.inbox.all().filter(q).order_by('-datetime_received')[:1]:
    if str(item.datetime_received) != file_datetime_received: # check that the new letter has a greater date than the previous one
        for attachment in item.attachments:
            if isinstance(attachment, FileAttachment):
                local_path = os.path.join(path_attch, attachment.name)
                with open(local_path, 'wb') as f:
                    f.write(attachment.content)
                print('Сохраняем вложение в ', local_path)
    else: 
        break
```

**Reply by ecederstrand** [reply, marked as answer]
It's a lot easier to use the stored datetime value if you convert it to a datetime when you read it, instead of using it as a string in your condition:

```python
with open("somefile.txt", "w") as f:
    f.write(some_date.isoformat())

with open("somefile.txt", "r") as f:
    some_date = datetime.fromisoformat(f.read())
```

**Reply by bamper197801** [reply]
This solution worked for me. Thank you.
